core/types/rlp_backwards_compat.libevm_test.go

```diff
@@ -106,3 +110,211 @@ func testHeaderRLPBackwardsCompatibility(t *testing.T) {
+		seed := seed
+		t.Run(fmt.Sprintf("seed_%d", seed), func(t *testing.T) {
+			t.Parallel()
+			testBodyRLPBackwardsCompatibility(t, seed)
+		})
+	}
+}
+
+//nolint:thelper
+func testBodyRLPBackwardsCompatibility(t *testing.T, seed uint64) {
+	rng := ethtest.NewPseudoRand(seed)
+
+	body := &Body{
+		// As this is a round-trip test, the slices MUST NOT be nil as the rlp
+		// package will always set them.
+		Transactions: []*Transaction{},
+		Uncles:       []*Header{},
+	}
+	for i, n := 0, rng.Intn(5); i < n; i++ {
+		tx := NewTx(&LegacyTx{
+			Nonce: rng.Uint64(),
+		})
+		body.Transactions = append(body.Transactions, tx)
+	}
+	for i, n := 0, rng.Intn(5); i < n; i++ {
+		body.Uncles = append(body.Uncles, &Header{
+			ParentHash: rng.Hash(),
+		})
+	}
+
+	var withdrawals Withdrawals
+	for i, n := 0, rng.Intn(4)+1; i < n; i++ {
+		withdrawals = append(withdrawals, &Withdrawal{
+			Index: rng.Uint64(),
+		})
+	}
+
+	tests := []struct {
+		name        string
+		withdrawals []*Withdrawal
+	}{
+		{
+			name: "with_nil_Withdrawals",
+		},
+		{
+			name: "with_empty_Withdrawals",
+			// As the struct field has an `rlp:"optional"` tag, we need to
+			// differentiate between nil (nothing added to RLP) and an empty
+			// slice (empty list added to RLP).
+			withdrawals: []*Withdrawal{},
+		},
+		{
+			name:        "with_Withdrawals",
+			withdrawals: withdrawals,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			body.Withdrawals = tt.withdrawals
+
+			// The original [Body] doesn't implement [rlp.Encoder] nor
+			// [rlp.Decoder] so we can use a methodless equivalent as the gold
+			// standard.
+			type withoutMethods Body
+			wantRLP, err := rlp.EncodeToBytes((*withoutMethods)(body))
+			require.NoErrorf(t, err, "rlp.EncodeToBytes([%T with methods stripped])", body)
+
+			t.Run("Encode", func(t *testing.T) {
+				got, err := rlp.EncodeToBytes(body)
+				require.NoErrorf(t, err, "rlp.EncodeToBytes(%T)", body)
+				assert.Equalf(t, wantRLP, got, "rlp.EncodeToBytes(%T)", body)
+			})
+
+			t.Run("Decode", func(t *testing.T) {
+				got := new(Body)
+				err := rlp.DecodeBytes(wantRLP, got)
+				require.NoErrorf(t, err, "rlp.DecodeBytes(..., %T)", got)
+
+				opts := cmp.Options{
+					cmpeth.CompareHeadersByHash(),
+					cmpeth.CompareTransactionsByBinary(t),
+				}
+				if diff := cmp.Diff(body, got, opts); diff != "" {
+					t.Errorf("rlp.DecodeBytes(rlp.EncodeToBytes(%T)) diff (-want +got):\n%s", body, diff)
+				}
+			})
+		})
+	}
+}
+
+// cChainBodyExtras carries the same additional fields as the ava-labs/coreth
+// [Body] and implements [BodyHooks] to achieve equivalent RLP {en,de}coding.
+type cChainBodyExtras struct {
+	Version uint32
+	ExtData *[]byte
+}
+
+var _ BodyHooks = (*cChainBodyExtras)(nil)
+
+func (e *cChainBodyExtras) AppendRLPFields(b rlp.EncoderBuffer, _ bool) error {
+	b.WriteUint64(uint64(e.Version))
+	if e.ExtData != nil {
+		b.WriteString(string(*e.ExtData))
+	} else {
+		b.WriteString("")
```

Comment:
nit I think this should work too

```suggestion
		b.WriteBytes(*e.ExtData)
	} else {
		b.WriteBytes([]byte{})
```

Reply:
Thanks; dunno how I missed that method! I've done it slightly DRYer but with `WriteBytes()`.